spec/contracts.dd

```diff
@@ -59,82 +61,106 @@ $(H2 $(LNAME2 pre_post_contracts, Pre and Post Contracts))
+        code.)
+
+        $(P The expression form is:)
+------
+in(expression)
+in(expression, "failure string")
+out(identifier; expression)
+out(identifier; expression, "failure string")
+out(; expression)
+out(; expression, "failure string")
+{
+    ...function body...
+}
+------
+        $(P The block statement form is:)
 ------
 in
 {
     ...contract preconditions...
 }
-out (result)
+out
+{
+    ...contract postconditions...
+}
+out (identifier)
 {
     ...contract postconditions...
 }
 do
 {
-    ...code...
+    ...function body...
 }
 ------
 
-    $(P Since $(LINK2 https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1003.md, DIP1003 has been applied)
-    the actual function body starts with $(D do).
-    In the past, $(D body) was used, and could still be encountered in old code bases.
-    In the long term, $(D body) may be deprecated, but for now it's allowed
-    as a keyword in this context and as an identifier in the function body,
-    although $(D do) is preferred.)
+        $(P The optional identifier in either type of `out` contract is set to the return value
+        of the function.)
 
         $(P By definition, if a pre contract fails, then the function received bad
-        parameters. An `AssertError` is thrown. If a post contract fails,
-        then there is a bug in the function. An `AssertError` is thrown.)
-
-        $(P Either the `in` or the `out` clause can be omitted.
-        If the `out` contract is for a function
-        body, the given `result` variable is declared and assigned the return
-        value of the function. For example, let's implement a square root function:)
-
-------
-long square_root(long x)
-in
-{
-    assert(x >= 0);
-}
-out (result)
-{
-    assert(result ^^ 2 <= x && (result + 1) ^^ 2 > x);
-}
-do
-{
-    return cast(long)std.math.sqrt(cast(real)x);
-}
-------
-        $(P The `assert`s in the `in` and `out` bodies are called $(I contracts).
-        Any other D
-        statement or expression is allowed in the bodies, but it is important
+        parameters. If a post contract fails, then there is a bug in the function. In either case,
+        an `assert` statement within the corresponding `in` or `out` block will throw an
+        `AssertError`.)
+
+        $(P Notice that the keyword `do` can be used to announce the function body. While any number
+        of pre or post contracts of any form may follow each other, `do` is
+        required only when the last contract before the body is a block statement.
+        (Before the acceptance of
+        $(LINK2 https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1003.md, DIP1003),
+        the keyword $(D body) was required instead of `do`, and may still be encountered in
+        old code bases. In the long term, $(D body) may be deprecated, but for now it's allowed both
+        as a keyword in this context and as an identifier elsewhere (although $(D do) is preferred).) )
```

Comment:
Is it just me, or there is one more `)` here than there should be?

Reply:
I put the parens around "(Before the acceptance of... elsewhere (although `do` is preferred).) )"
The last parens are to close the paragraph $(P ...). I'll remove the "although `do`"  set of parens.
This file did pass DAutoTest before the grammar commits, so it's not the cause of the errors.

Reply:
vim says the parens are legit.